service/worker/deployment/workflow.go

```diff
@@ -0,0 +1,174 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2024 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deployment
+
+import (
+	"time"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	sdkclient "go.temporal.io/sdk/client"
+	sdklog "go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/temporal"
+
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	TaskQueue struct {
+		Name          string
+		TaskQueueType enumspb.TaskQueueType
+	}
+	Deployment struct {
+		Namespace      string
+		DeploymentName string
+		TaskQueue      []*TaskQueue // All the task queues associated with this buildID/deployment
+		BuildID        string
+	}
+	DeploymentName struct {
+		Name           string
+		CurrentBuildID string // denotes the current "default" build-ID of a deploymentName
+	}
+
+	// DeploymentWorkflowRunner holds the local state while running a deployment workflow
+	DeploymentWorkflowRunner struct {
+		deployment Deployment
+		ctx        workflow.Context
+		a          *activities
+		logger     sdklog.Logger
+		metrics    sdkclient.MetricsHandler
+	}
+)
+
+var (
+	defaultActivityOptions = workflow.ActivityOptions{
+		ScheduleToCloseTimeout: 1 * time.Hour,
+		StartToCloseTimeout:    30 * time.Second,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: 1 * time.Second,
+			MaximumInterval: 60 * time.Second,
+		},
+	}
+)
+
+const continueAsNewThreshold = 1000
+
+func DeploymentWorkflow(ctx workflow.Context, deploymentWorkflowArgs Deployment) error {
+	deploymentWorkflowRunner := &DeploymentWorkflowRunner{
```

Comment:
If the workflow is simple enough, you don't need to use this "struct with a .run() method" structure. That's just for the schedule workflow since it's so complicated. Most workflows don't (and shouldn't) use it.

Reply:
I thought it was pretty neat structure given that the workflow's local state (which arguably is the most important entity for this workflow), a logger, metricsHandler can all be attached to a structure. In this current iteration of review, this current code shall be removed (explained offline as to why I am doing this) but we can come back to this as and when I post this back up again. ( I shall also give it more thought to incorporate your feedback)

Reply:
It's fine, it's just slightly contrary to best practices for using the Go SDK that we give out. The argument is that it makes it too easy to introduce nondeterministic dependencies. I'm fine either way.